plugins/kotlin/fir/src/org/jetbrains/kotlin/idea/completion/contributors/FirDeclarationFromUnresolvedNameContributor.kt

```diff
@@ -0,0 +1,149 @@
+        currentDeclarationInFakeFile: KtNamedDeclaration,
+        unresolvedRef: KtNameReferenceExpression
+    ) {
+        // In a block, references must be after the declaration. Therefore, we only offer completion for unresolved names that appear after
+        // the current cursor position.
+        if (referenceScope is KtBlockExpression && unresolvedRef.startOffset < parameters.offset) {
+            return
+        }
+        val name = unresolvedRef.getReferencedName()
+        if (!prefixMatcher.prefixMatches(name)) return
+        if (!shouldOfferCompletion(unresolvedRef, currentDeclarationInFakeFile)) return
+        if (unresolvedRef.reference?.resolve() == null) {
+            val lookupElement = LookupElementBuilder.create(name).suppressAutoInsertion()
+                .also { it.priority = ItemPriority.FROM_UNRESOLVED_NAME_SUGGESTION }
+            sink.addElement(lookupElement)
+        }
+    }
+
+    private fun KtAnalysisSession.shouldOfferCompletion(
+        unresolvedRef: KtNameReferenceExpression,
+        currentDeclarationInFakeFile: KtNamedDeclaration
+    ): Boolean {
+        val refExprParent = unresolvedRef.parent
+        val refExprIsCallRef = refExprParent is KtCallableReferenceExpression
+        val refExprIsCall = refExprParent is KtCallExpression
+        val refExprIsType = refExprParent is KtUserType
+        val receiver = if (refExprIsCall) refExprParent.getReceiverForSelector() else unresolvedRef.getReceiverForSelector()
+        return when (val symbol = (basicContext.toFromOriginalFileMapper.toOriginalFile(currentDeclarationInFakeFile)
+        // fallback to the symbol in the fake file if usr has not typed anything
+            ?: currentDeclarationInFakeFile).getSymbol()) {
+            is KtCallableSymbol -> {
+                if (refExprIsType) return false
+                // If current declaration is a function, we only offer names of unresolved function calls. For property declarations, we
+                // always offer an unresolved usage because a property may be called if the object has an `invoke` method.
+                if (!refExprIsCallRef && !refExprIsCall && symbol is KtFunctionLikeSymbol) return false
+                if (receiver != null) {
+                    val actualReceiverType = receiver.getKtType() ?: return false
+                    val expectedReceiverType = getReceiverType(symbol) ?: return false
+                    !(actualReceiverType isNotSubTypeOf expectedReceiverType)
+                } else {
+                    // If there is no explicit receiver at call-site, we check if any implicit receiver at call-site matches the extension
+                    // receiver type for the current declared symbol
+                    val extensionReceiverType = symbol.receiverType?.type
+                    if (extensionReceiverType != null) {
+                        !getImplicitReceiverTypesAtPosition(unresolvedRef).none { it isSubTypeOf extensionReceiverType }
```

Comment:
`!list.none(p) <=> list.any(p)` it is just more readable without double negation

Reply:
Thanks!